---

output: 
    html_document:
        highlight: default
        theme: paper
        code_folding: hide
        df_print: tibble
        toc: true
        toc_depth: 3
        toc_float: true
        

---
    
    

```{r setup, include=FALSE}
knitr::opts_chunk$set(
    tidy = TRUE,
    tidy.opts = list(width.cutoff = 120),
    message = FALSE,
    warning = FALSE
)

# if (length(.libPaths()) > 2){
#     .libPaths(.libPaths()[2:3])
# }

# myPaths <- .libPaths()
# myNewPaths <- c("/camp/stp/babs/working/boeings/Projects/pachnisv/song.chng/195B_RNAseq_enteric_neurons_DIV_timecourse/basedata", myPaths)
# .libPaths(myNewPaths)
# library(renv)
# .libPaths(myPaths)

#renv::init()
```



```{r populate_meta_data_database, eval=FALSE, echo=F, results=F}


###############################################################################
## Upload Anna's samples                                                     ##


# Load Seurat Object from rds file

#rdsFN <- "/camp/lab/pachnisv/working/SingleCellData/Internal/SmallIntestineDevelopmentalData/SmallIntestineDevelopmentalData.rds"
#rdsFN <- "/camp/lab/pachnisv/working/SingleCellData/Internal/MouseGliaIfngrKO/MouseGliaIfngrKO.rds"
rdsFN <- "/camp/lab/pachnisv/working/SingleCellData/Internal/EGCsHPoly/EGCsHPoly.rds"
#rdsFN <- "/camp/lab/pachnisv/working/SingleCellData/Internal/InVitroData/InVitroData.rds"
#rdsFN <- "/camp/lab/pachnisv/working/SingleCellData/Internal/InVitroData/InVitroFoxd3RAData.rds"

## Object received from Michael ##
#rdsFN <- "/camp/lab/pachnisv/working/Michael/FranzeColouredSeuratObject/ColoredSeuratObject.rds"
## name to be used PubMouseGliaIfngrKO

h <- unlist(strsplit(rdsFN, "\\/"))[length(unlist(strsplit(rdsFN, "\\/")))]
project_id <- gsub(".rds", "", h)
## Done                                                                      ##
###############################################################################


```

What you need before you start:
  A MySQL database that can be accessed remotely
  A user for this database that is authorized to insert, select and delete. 
  
```{r main_bit, echo=F, eval=T, warning=FALSE, results="asis"}


###############################################################################
## Create folder structure                                                   ##

## Create user in db
if (exists("host")){
    if (host == "10.27.241.82"){
        serverString <- "shiny-bioinformatics.crick.ac.uk"
        shinyBasePath <- "/camp/stp/babs/www/shiny/boeings/external/"  
        #FvDir <- paste0("/camp/stp/babs/www/boeings/bioLOGIC/data/", project_id)
    } else if (host == "10.27.241.234"){
        serverString <- "shiny.thecrick.org"
        shinyBasePath <- "/camp/stp/babs/www/shiny/boeings/"
        #FvDir <- paste0("/camp/stp/babs/www/boeings/bioLOGIC/data/", project_id)
    }
} else {
    serverString <- "shiny-bioinformatics.crick.ac.uk"
    shinyBasePath <- "./"
    #FvDir <- paste0("./",project_id,"_app/data/", project_id)
    #stop("no server specified")
}  

## Remove if it exists
shinyProjectPath <-paste0(
  shinyBasePath, 
  project_id, "_app"
)
if (!dir.exists(shinyProjectPath)){
  dir.create(shinyProjectPath)
}

shinyLocalDbPath <-paste0(shinyProjectPath, "/data/")
if (!dir.exists(shinyLocalDbPath)){
  dir.create(shinyDataPath, recursive = T)
}

shinyDataPath <-paste0(shinyProjectPath, "/data/connect/")
if (!dir.exists(shinyDataPath)){
  dir.create(shinyDataPath, recursive = T)
}
shinyParamPath <-paste0(shinyProjectPath, "/data/parameters/")
if (!dir.exists(shinyParamPath)){
  dir.create(shinyParamPath)
}

## Done                                                                      ##
###############################################################################

###############################################################################
## Set Project Parameters                                                    ##

#project_id <- "InVitroFoxd3RAData"
#project_id <- "PubMouseGliaIfngrKO"

# host = "10.27.241.82"
# db.user = "boeingS"
# password = db.pwd
primDataDB = paste0(shinyLocalDbPath, "appDataDb")


PCAdbTableName <- paste0(
    project_id, 
    "_PCA"
)


##                                                                           ##
###############################################################################
```

```{r main_bit, echo=F, eval=T, warning=FALSE, results="asis"}

###############################################################################
##
devtools::install_github("decusInLabore/biologicSeqTools")


##
###############################################################################

###############################################################################
## Step 1: Load Seurat object                                                ##

OsC <- readRDS(rdsFN)
names(OsC@meta.data) <- gsub("\\.", "_", names(OsC@meta.data))
OsC@meta.data[["clusterName"]] <- OsC@meta.data$cluster
OsC@meta.data[["sampleName"]] <- OsC@meta.data$orig_ident
OsC@meta.data[["all"]] <- "all"
  

##                                                                           ##
###############################################################################

###############################################################################
## Create Expr table                                                         ##


## Running this function may take a minute or two, depending on the number of cells in your dataset

dfExpr <-  biologicSeqTools::createDfExpr(
    obj = OsC,
    assay = "RNA",
    #slot = "data",
    geneSel = NULL
) 

## In Sqlite version load via function ##



## Adding gene_ids 

library("dplyr")

dfIDTable <- dfExpr %>% 
    select(gene) %>% 
    distinct() %>% 
    mutate(gene_id = row_number())


## Upload expression table to database 

print(paste0("Database to be used: ", primDataDB))
print(paste0("Database table name to be used: ", paste0(project_id, "_geneID_tb")))

biologicSeqTools::upload.datatable.to.database(
    #host = host,
    #user = db.user,
    #password = db.pwd,
    prim.data.db = primDataDB,
    dbTableName = paste0(project_id, "_geneID_tb"),
    df.data = dfIDTable,
    db.col.parameter.list = list(
        "BIGINT(8) NULL DEFAULT NULL" = c("row_names"),
        "VARCHAR(100) CHARACTER SET utf8 COLLATE utf8_general_ci" = c("gene"),
        "INT(8) NULL DEFAULT NULL" = c("gene_id")
    ),
    new.table = T,
    cols2Index = c("gene"),
    mode = "SQLite"  # Options: "MySQL" and "SQLite"
)


## Rearrange expression talbe
## This step may take a couple of minutes in a large dataset
dfExpr <- dfExpr %>% 
    rename(condition = cellID)  %>%  
    mutate(lg10Expr = round(lg10Expr, 3)) %>% 
    arrange(gene) 




## Save as temp file for more efficient database upload
#tempFN <- paste0("temp.",project_id,".csv")
#write.csv(dfExpr, tempFN, row.names = F)

# You may want to remove dfExpr to save on memory space
# rm(dfExpr)

expDbTable <- paste0(project_id, "_gene_expr_tb")


## Upload expression table to database 

print(paste0("Database to be used: ", primDataDB))
print(paste0("Database table name to be used: ", expDbTable))

colCatList <- biologicSeqTools::inferDBcategories(dfExpr)

biologicSeqTools::upload.datatable.to.database(
    #host = host,
    #user = db.user,
    #password = db.pwd,
    prim.data.db = primDataDB,
    dbTableName = expDbTable,
    df.data = dfExpr,
    db.col.parameter.list = colCatList,
    new.table = T,
    cols2Index = c("gene"),
    #indexName = c("idx_gene_exp"),
    mode = "SQLite"  # Options: "MySQL" and "SQLite"
)


###############################################################################
## Create Metadata table                                                     ##

dfCoord <- biologicSeqTools::createDfCoord(OsC)


dupTest <- duplicated(toupper(names(dfCoord)))

if (sum(dupTest) > 0 ){
  names(dfCoord)[duplicated(toupper(names(dfCoord)))] <- paste0( names(dfCoord)[duplicated(toupper(names(dfCoord)))], "_B")
}

dfCoord <- dfCoord [,!(duplicated(names(dfCoord)))]

dfdbTable <- dfCoord

pos <- grep("integrated_", names(dfdbTable))
if (length(pos) > 0){
  dfdbTable <- dfdbTable[,-pos]
}
names(dfdbTable) <- gsub("\\.", "_", names(dfdbTable))

write.table(
  dfdbTable,
  "temp.txt",
  row.names = F,
  sep = "\t"
)

dfdbTable <- read.delim(
  "temp.txt",
  sep = "\t",
  stringsAsFactors = F
)

if (file.exists("temp.txt")){
  unlink("temp.txt")
}

dfdbTable[is.na(dfdbTable)] <- ""



columnDBcategoryList <- biologicSeqTools::inferDBcategories(dfData=dfdbTable)


biologicSeqTools::upload.datatable.to.database(
  host = host,
  user = db.user,
  password = db.pwd,
  prim.data.db = primDataDB,
  dbTableName = PCAdbTableName,
  df.data = dfdbTable,
  db.col.parameter.list = columnDBcategoryList,
  new.table = TRUE,
   mode = "SQLite"
)
biologicSeqTools::killDbConnections()
##                                                                           ##
###############################################################################

###############################################################################
## Menu Options                                                              ##
paramList <- biologicSeqTools::scanObjParams(OsC)

## Done                                                                      ##
###############################################################################

###############################################################################
## Select default splitBy options                                            ##

## Default all non-numeric columns
numCols <- unlist(lapply(dfCoord, is.numeric))
chrCols <- unlist(lapply(dfCoord, is.character))

splitOptions <- names(dfCoord)[chrCols]

splitOptions <- c(splitOptions[splitOptions == "all"], sort(splitOptions[splitOptions != "all"]))

names(splitOptions) <- splitOptions
names(splitOptions) <- gsub("all", "None", names(splitOptions))

firstup <- function(x) {
  substr(x, 1, 1) <- toupper(substr(x, 1, 1))
  x
}

names(splitOptions) <- sapply(names(splitOptions), function(x) firstup(x))


paramList$splitPlotsBy <- NULL

paramList$splitPlotsBy <- splitOptions

## Make None the first option

##                                                                           ##
###############################################################################

###############################################################################
## Select colorBy options                                                    ##


## Default all non-numeric columns
numCols <- unlist(lapply(dfCoord, is.numeric))
chrCols <- unlist(lapply(dfCoord, is.character))

colorByOptions <- sort(names(dfCoord))

## Columns not to show in the color selection
rmVec <- c(
    grep("^UMAP", toupper(colorByOptions)),
    grep("^TSNE", toupper(colorByOptions)),
    grep("^PC", toupper(colorByOptions)),
    grep("orig.ident", toupper(colorByOptions)),
    grep("cellID", toupper(colorByOptions))
    
)

if (length(rmVec) > 0){
    colorByOptions <- colorByOptions[-rmVec]
}

colorByOptionsPart1 <- c(
    grep("clusterName", colorByOptions),
    grep("seurat_clusters", colorByOptions),
    grep("sampleName", colorByOptions)
)

if (length(colorByOptionsPart1) > 0){
    colorByOptionsPart2 <- colorByOptions[-colorByOptionsPart1]
}

colorDisplayOptions <- c(
    "lg10Expr",
    colorByOptions[colorByOptionsPart1],
    sort(colorByOptionsPart2)
)

names(colorDisplayOptions) <- colorDisplayOptions
names(colorDisplayOptions) <- gsub("all", "Uniform", names(colorDisplayOptions))
names(colorDisplayOptions) <- gsub("lg10Expr", "log10 Expr", names(colorDisplayOptions))

firstup <- function(x) {
  substr(x, 1, 1) <- toupper(substr(x, 1, 1))
  x
}

names(colorDisplayOptions) <- sapply(names(colorDisplayOptions), function(x) firstup(x))
names(colorDisplayOptions) <- gsub("_", " ", names(colorDisplayOptions))


paramList$colorPlotsBy <- NULL

paramList$colorPlotsBy <- colorDisplayOptions

## Order Options ##

##                                                                           ##
###############################################################################

###############################################################################
## Edit x-axis                                                               ##

## Done                                                                      ##
###############################################################################

###############################################################################
## Edit y-axis                                                               ##

## Done                                                                      ##
###############################################################################


###############################################################################
## Helper function                                                           ##

createListChunk <- function(
  optionName = "Choose_an_X-axis",
  pList = paramList,
  listItem = "x_axis"
){
  dfTemp <- data.frame(
      menuName = rep( optionName,length(pList[listItem])), 
      displayName = names(pList[listItem]), 
      colSel=pList[[listItem]],
      displayOrder = 1:length(pList[[listItem]])
  )
  return(dfTemp)
}

##                                                                           ##
###############################################################################

dfTemp <- createListChunk(
    optionName = "Choose_an_X-axis",
    pList = paramList,
    listItem = "x_axis"
)

df <- dfTemp

dfTemp <- createListChunk(
    optionName = "Choose_an_Y-axis",
    pList = paramList,
    listItem = "y_axis"
)

df <- rbind(
    dfTemp,
    df
)

dfTemp <- createListChunk(
    optionName = "Split_Plots_By",
    pList = paramList,
    listItem = "splitPlotsBy"
)

df <- rbind(
    dfTemp,
    df
)

dfTemp <- createListChunk(
    optionName = "Color_Plots_By",
    pList = paramList,
    listItem = "colorPlotsBy"
)

df <- rbind(
    dfTemp,
    df
)

write.table(
    df,
    paste0(shinyParamPath, "menuParameters.txt"),
    sep = "\t",
    row.names = F
)

##                                                                           ##
###############################################################################

###############################################################################
## Make color table                                                          ##

## Done                                                                      ##
###############################################################################

## Done 
###############################################################################


aFN <- paste0(shinyDataPath, "db.txt")

if (file.exists(aFN)){
  df <- read.delim(aFN, header = T, sep="\t", stringsAsFactors = F)
  sPwd <- as.vector(df$id2)
  sUser <- as.vector(df$id)
} else {
    sUser <- substr(paste0(project_id, "_scData"), 1, 15)
    sPwd <-c(2:9,letters,LETTERS)
    sPwd <- paste(sample(sPwd,8),collapse="")
}


query <- as.vector(NULL, mode = "character")






# 
# GRANT SELECT on csl_data.p316_rna_seq_table TO csl316data@'shiny-bioinformatics.crick.ac.uk';
## Create relevant directory ##
## Create log-in file ##


 


if (!(default %in% dfExpr$gene)){
    stop("Set default gene correctly")
    
}

## Maintain old password if exists
# FN <- paste0(shinyDataPath, "db.txt")
# if (file.exists(FN)){
#     dfCont <- read.delim(
#         FN, 
#         header = T,
#         sep = "\t",
#         stringsAsFactors = F
#     )
#     dfID$id2 <- dfCont$id2
# }
#write.table(dfID, FN, row.names = F, sep="\t")
## Copy rest ##



########################
## Temp section
devtools::install_github("decusInLabore/biologicViewerSC")
library(biologicViewerSC)

pathCP <- system.file("scAppTemplate",package="biologicViewerSC")

##
##########################

cpString1 <- paste0("cp -r ",pathCP,"/* ", shinyProjectPath)


#cpString1 <- paste0("cp assets/template_sc_app/ui.r ", shinyProjectPath)
system(cpString1)



dfID <- data.frame(
    dataMode = "SQLite",
    url = "",
    id = "",
    id2 = "",
    db =  gsub(paste0("./", project_id, "_app/"), "",primDataDB),
    coordTb = PCAdbTableName,
    exprTb = paste0(project_id, "_gene_expr_tb"),
    geneTb = paste0(project_id, "_geneID_tb"),
    default = as.vector(dfExpr[1,"gene"])
)

FN <- paste0(shinyDataPath, "db.txt")
write.table(dfID, FN, row.names = F, sep="\t")
# cpString2 <- paste0("cp assets/template_sc_app/server.r ", shinyProjectPath)
# system(cpString2)
## Done creating app ##


##                                                                           ##
###############################################################################



###############################################################################
## Customise app                                                             ##

params <- biologicSeqTools::scanObjParams(OsC)

## Done                                                                      ##
###############################################################################
```




## Documentation
```{r documentation, eval=TRUE, echo=T, results=T}
sessionInfo()
```

---
title: Create Single Cell Viewer connected to a MySQL database
subtitle:  SeuratObject to Database
author:
    - Bioinformatics: Stefan Boeing^[The Francis Crick Institute, stefan.boeing@crick.ac.uk]
date: 'Compiled: `r format(Sys.Date(), "%B %d, %Y")`'

abstract: |
    This recepie will create a single-cell viewer from your Seurat object
    
css: `system.file("src/assets/style/style.css" ,package="biologicViewerSC)`    

---